### PR TITLE
MAINT: skip setting rpath if it's not needed

### LINKED ISSUE #125: Only use and depend on `patchelf` if it is actually needed
Sage ran into an issue where `patchelf` is built from source and break with GCC 6.3 on Debian when trying to use `meson-python` with SciPy (see https://trac.sagemath.org/ticket/34081, Cc @mkoeppe). This is a little unfortunate, because `patchelf` is not actually needed. What it's used for here is to add a non-existing folder to the RPATH - which is not useful:

```python
        # fix file
        if platform.system() == 'Linux':
            # add .mesonpy.libs to the RPATH of ELF files
            if self._is_elf(os.fspath(origin)):
                # copy ELF to our working directory to avoid Meson having to regenerate the file
                new_origin = self._libs_build_dir / pathlib.Path(origin).relative_to(self._project._build_dir)
                os.makedirs(new_origin.parent, exist_ok=True)
                shutil.copy2(origin, new_origin)
                origin = new_origin
                # add our in-wheel libs folder to the RPATH
                elf = mesonpy._elf.ELF(origin)
                libdir_path = f'$ORIGIN/{os.path.relpath(f".{self._project.name}.mesonpy.libs", destination.parent)}'
                if libdir_path not in elf.rpath:
                    elf.rpath = [*elf.rpath, libdir_path]
```

Of course an issue is that `meson-python` cannot currently know when setting up the build environment whether the project has standalone compiled libraries that are going to be included in `.mesonpy.libs`. This would be easy to make controllable via a boolean flag in `pyproject.toml` though - and then the vast majority of Python packages that do not have such libraries don't have to deal with the unnecessary `patchelf` dependency.

It seems to me like it's time for a `[meson-python]` table in `pyproject.toml`, with at least two entries: one for "has standalone libraries" and one for "pure" to avoid the fragile heuristics around that.

### PR DESCRIPTION
**Pull request by FFY00**
I also fixed `ELF.rpath`, which is needed for the tests, and added a test for setting the rpath in the first place in this PR.

Fixes #125

### CODE REVIEW

File: tests/test_wheel.py [hunk 1]
```diff
@@ -150,3 +154,21 @@ def test_executable_bit(wheel_executable_bit):
             assert executable_bit, f'{info.filename} should have the executable bit set!'
         else:
             assert not executable_bit, f'{info.filename} should not have the executable bit set!'
+
+
+def test_rpath(wheel_link_against_local_lib, tmpdir):
```
Comment (rgommers) [inline tests/test_wheel.py:159]: These added test should should be skipped unless the platform is `linux`.

File: mesonpy/_elf.py [hunk 2]
```diff
@@ -21,9 +21,10 @@ def _patchelf(self, *args: str) -> str:
 
     @property
     def rpath(self) -> Collection[str]:
-        if not self._rpath:
-            self._rpath = self._patchelf('--print-rpath').strip().split(';')
-        return self._rpath
+        if self._rpath is None:
+            rpath = self._patchelf('--print-rpath').strip()
```
Comment (rgommers) [inline mesonpy/_elf.py:25]: This doesn't work portably it looks like. For example [in this CI log](https://github.com/FFY00/meson-python/actions/runs/3049095667/jobs/4914826119#step:7:340):
```
 E               subprocess.CalledProcessError: Command '['patchelf', '--print-rpath', '/tmp/pytest-of-runner/pytest-0/test_uneeded_rpath0/purelib_and_platlib-1.0.0.data/platlib/plat.{EXT_SUFFIX}']' returned non-zero exit status 1.
```

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @FFY00. There are some issues shown by the CI results. The code change looks like an improvement, and will solve the most important part of gh-125. It looks like it won't yet solve the unnecessary dependency issue. That can be left for later if you prefer.

**Comment by rgommers** [comment]
Ugh, this still depends on other fixes:
```
patchelf: getting info about '/tmp/pytest-of-runner/pytest-0/test_unneeded_r"
 b"path0/plat.cpython-39-x86_64-linux-gnu.so': No such file or directory
```
That's because the test package it's built on is still xfailed:
```python
@pytest.mark.xfail(reason='Meson bug')
def test_purelib_and_platlib(wheel_purelib_and_platlib):
    artifact = wheel.wheelfile.WheelFile(wheel_purelib_and_platlib)

    expecting = {
        f'plat{EXT_SUFFIX}',
```
This passes locally with Meson master. After the other open PR for tags is merged, it should pass with released Meson as well.

**Comment by rgommers** [comment]
You just lost the commits that made this PR pass CI. I'll re-add them.

**Comment by FFY00** [comment]
Sorry :sweat_smile:

**Comment by rgommers** [comment]
I think that's it, but hard to be sure since I had to manually re-apply fixes. Let's first ensure CI is green, and then if desired at the end rewrite history to the desired commit granularity.

**Comment by FFY00** [comment]
I have squashed the fix with the affected commit, the tests seem to be passing, are you okay with me pushing?

**Comment by rgommers** [comment]
Sure, go ahead. CI is green (codecov isn't relevant, and Sage isn't currently testing the right thing). So rewrite history as you wish and then merge it I'd say.

**Pull request merged**